File: typescript/src/core/modules/pcustomer-management/functions/get-pcustomer.ts
import { AppError, isStandardError } from "../../../error";
import { PCustomerDAO } from "../data-model/dao";
import { initializeInternalContext } from "../../../context/internal-context";

// Import types
import type { RuntimeContext } from "../../../context/runtime-context";
import type { InternalContext } from "../../../context/internal-context";
import type { TFindPCustomerParams } from "../data-model/type";

/**
 * Lấy một người dùng ở trong cơ sở dữ liệu, theo một số điều kiện.
 *
 * @param ctx - runtime context
 */
export async function getCustomer(ctx: RuntimeContext) {
  try {
    const params = await ctx.getParams<{ id: string }>();

    const pcustomerDao = new PCustomerDAO();
    const internalCtx =
      initializeInternalContext() as InternalContext<TFindPCustomerParams>;

    internalCtx.params = {
      query: { id: params.id },
    };
    internalCtx.options!.canCatchError = true;

    const result = await pcustomerDao.getPCustomer(internalCtx);

    return result;
  } catch (error) {
    if (isStandardError(error)) return error;

    const err = new AppError("Cannot get a potential customer");
    err.asHTTPError("InternalServerError");
    return err;
  }
}

File: python/src/core/modules/pcustomer_management/functions/get_pcustomer.py
# Import from core
from core.context.runtime_context import RuntimeContext
from core.context.internal_context.main import initialize_internal_context
from core.error import AppError, is_standard_error
from core.modules.pcustomer_management.data_model.dao import PCustomerDAO


async def get_pcustomer(ctx: RuntimeContext):
    """
    Lấy thông tin của một khách hàng ở trong hệ thống.

    Args:
        ctx (RuntimeContext): Ngữ cảnh runtime chứa thông tin request.

    Returns:
        dict | AppError: Thông tin khách hàng hoặc lỗi chuẩn hóa.
    """
    try:
        params = ctx.get_params()

        pcustomer_dao = PCustomerDAO()
        internal_ctx = initialize_internal_context()

        internal_ctx.params = {"query": {"id": params.get("id")}}
        internal_ctx.options["can_catch_error"] = True

        result = await pcustomer_dao.get_pcustomer(internal_ctx)
        return result

    except Exception as error:
        if is_standard_error(error):
            return error

        err = AppError("Cannot get potential customer")
        err.as_http_error("InternalServerError")
        return err
